pdns/opensslsigners.cc

```diff
@@ -627,6 +660,47 @@ void OpenSSLECDSADNSCryptoKeyEngine::create(unsigned int bits)
   }
 }
 
+auto OpenSSLECDSADNSCryptoKeyEngine::createFromPEMFile(DNSKEYRecordContent& drc, const string& filename, std::FILE* fp) -> void
+{
+  drc.d_algorithm = d_algorithm;
+  d_eckey = std::unique_ptr<EC_KEY, decltype(&EC_KEY_free)>(PEM_read_ECPrivateKey(fp, nullptr, nullptr, nullptr), &EC_KEY_free);
+  if (d_eckey == nullptr) {
+    throw runtime_error(getName() + ": Failed to read private key from PEM file `" + filename + "`");
+  }
+
+  int ret = EC_KEY_set_group(d_eckey.get(), d_ecgroup.get());
+  if (ret != 1) {
+    throw runtime_error(getName() + " setting key group failed");
+  }
+
+  const BIGNUM* privateKeyBN = EC_KEY_get0_private_key(d_eckey.get());
+
+  auto pub_key = std::unique_ptr<EC_POINT, void (*)(EC_POINT*)>(EC_POINT_new(d_ecgroup.get()), EC_POINT_free);
+  if (!pub_key) {
+    throw runtime_error(getName() + " allocation of public key point failed");
+  }
+
+  ret = EC_POINT_mul(d_ecgroup.get(), pub_key.get(), privateKeyBN, nullptr, nullptr, nullptr);
+  if (ret != 1) {
+    throw runtime_error(getName() + " computing public key from private failed");
+  }
+
+  ret = EC_KEY_set_public_key(d_eckey.get(), pub_key.get());
+  if (ret != 1) {
+    ERR_print_errors_fp(stderr);
+    throw runtime_error(getName() + " setting public key failed");
+  }
+
+  EC_KEY_set_asn1_flag(d_eckey.get(), OPENSSL_EC_NAMED_CURVE);
```

Comment:
Do you know when we actually need this? After reading https://wiki.openssl.org/index.php/Elliptic_Curve_Diffie_Hellman it's not clear to me if we need this only when we generated the key ourselves, when exporting it, or when loading it..

Reply:
I don't actually know except that "the key isn't going to be very useful without it". And having the named OID is nice, it helped us notice an inconsistency when testing with `openssl asn1parse`.

Is it harmful here in any way?

Reply:
No, I'm not advocating to remove it here. I was instead wondering if we need to add it in other places, now that we know it exists.